src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -0,0 +1,257 @@
+    } else if (len >= size) {
+      clone()
+    } else {
+      val array2 = ArrayDeque.alloc(len)
+      arrayCopy(dest = array2, srcStart = left, destStart = 0, maxItems = len)
+      new ArrayDeque(array2, 0, len)
+    }
+  }
+
+  override def sliding(window: Int, step: Int) = {
+    require(window >= 1 && step >= 1, s"size=$size and step=$step, but both must be positive")
+    (indices by step).iterator.map(i => slice(i, i + window))
+  }
+
+  override def grouped(n: Int) = sliding(n, n)
+
+  override def copyToArray[B >: A](dest: Array[B], destStart: Int, len: Int) =
+    arrayCopy(dest = dest, srcStart = 0, destStart = destStart, maxItems = len)
+
+  override def companion = ArrayDeque
+
+  override def result() = this
+
+  override def stringPrefix = "ArrayDeque"
+
+  override def toArray[B >: A: ClassTag] = {
+    val array2 = Array.ofDim[B](size)
+    arrayCopy(dest = array2, srcStart = 0, destStart = 0, maxItems = size)
+    array2
+  }
+
+  def arrayCopy(dest: Array[_], srcStart: Int, destStart: Int, maxItems: Int): Unit = {
+    checkIndex(destStart, dest)
+    checkIndex(srcStart)
+    val toCopy = maxItems min (size - srcStart) min (dest.length - destStart)
+    if (toCopy > 0) {
+      val startIdx = (start + srcStart) & mask
+      val block1 = toCopy min (array.length - startIdx)
+      Array.copy(src = array, srcPos = startIdx, dest = dest, destPos = destStart, length = block1)
+      if (block1 < toCopy) {
+        Array.copy(src = array, srcPos = 0, dest = dest, destPos = block1, length = toCopy - block1)
+      }
+    }
+  }
+
+  /**
+    * Trims the capacity of this CircularBuffer's instance to be the current size
+    */
+  def trimToSize(): Unit = resize(size - 1)
+
+  @inline private def checkIndex(idx: Int, seq: GenSeq[_] = this) = if(!seq.isDefinedAt(idx)) throw new IndexOutOfBoundsException(idx.toString)
+
+  @inline private def box(i: Int) = if (i <= 0) 0 else if (i >= size) size else i
+
+  @inline private def get(idx: Int) = array((start + idx) & mask)
+
+  @inline private def set(idx: Int, elem: AnyRef) = array((start + idx) & mask) = elem
+
+  @inline private def set(array: Array[AnyRef], start: Int, end: Int) = {
+    this.array = array
+    this.mask = array.length - 1
+    assert((array.length & mask) == 0, s"Array.length must be power of 2")
+    this.start = start
+    this.end = end
+  }
+
+  private def ensureCapacity() = if (size == array.length - 1) resize(array.length)
+
+  private def resize(len: Int) = {
+    val array2 = ArrayDeque.alloc(len)
+    arrayCopy(dest = array2, srcStart = 0, destStart = 0, maxItems = size)
+    set(array = array2, start = 0, end = size)
+  }
+}
+
+object ArrayDeque extends generic.SeqFactory[ArrayDeque] {
+  implicit def canBuildFrom[A]: generic.CanBuildFrom[Coll, A, ArrayDeque[A]] =
+    ReusableCBF.asInstanceOf[GenericCanBuildFrom[A]]
+
+  override def newBuilder[A]: mutable.Builder[A, ArrayDeque[A]] = new ArrayDeque[A]()
+
+  val defaultInitialSize = 8
```

Comment:
I'm referring to constants as the term is used in the [Scala style guide](http://docs.scala-lang.org/style/naming-conventions.html#constants-values-variable-and-methods):

> Constant names should be in upper camel case. Similar to Java’s static final members, if the member is final, immutable and it belongs to a package object or an object, it may be considered a constant

I don't have a really strong opinion on this, but I assumed that we would follow the official Scala style guide here. That seems to distinguish constants from simple values.

Also, I agree with @Ichoran the `final` modifier should be added.